--- code/src/molecule.h	2025-07-22 03:37:51.000000000 +0200
+++ code/src/molecule.h	2025-08-20 18:26:17.029593962 +0200
@@ -51,6 +51,7 @@
   int nspecialflag, specialflag;
   int shakeflag, shakeflagflag, shakeatomflag, shaketypeflag;
   int bodyflag, ibodyflag, dbodyflag;
+  int repatomflag;
 
   // 1 if attribute defined or computed, 0 if not
 
@@ -69,6 +70,7 @@
   double *radius;      // radius of each atom
   double *rmass;       // mass of each atom
   double **mu;         // dipole vector of each atom
+  int *rep_atom;       // mark representative atom
 
   int *num_bond;    // bonds, angles, dihedrals, impropers for each atom
   int **bond_type;
@@ -164,6 +166,7 @@
   void shakeatom_read(char *);
   void shaketype_read(char *);
   void body(int, int, char *);
+  void repatoms(char *);
 
   void initialize();
   void allocate();

--- code/src/molecule.cpp	2025-07-22 03:37:51.000000000 +0200
+++ code/src/molecule.cpp	2025-08-20 18:45:55.314702554 +0200
@@ -56,7 +56,7 @@
     special(nullptr), shake_flag(nullptr), shake_atom(nullptr), shake_type(nullptr),
     avec_body(nullptr), ibodyparams(nullptr), dbodyparams(nullptr), fragmentmask(nullptr),
     dx(nullptr), dxcom(nullptr), dxbody(nullptr), quat_external(nullptr), fp(nullptr),
-    count(nullptr)
+    count(nullptr), rep_atom(nullptr)
 {
   // parse args until reach unknown arg (next file)
 
@@ -166,7 +166,8 @@
     if (comm->me == 0) {
       fp = fopen(filename.c_str(), "r");
       if (fp == nullptr)
-        error->one(FLERR, fileiarg, "Cannot open molecule file {}: {}", filename, utils::getsyserror());
+        error->one(FLERR, fileiarg, "Cannot open molecule file {}: {}", filename,
+                   utils::getsyserror());
       try {
         // try to parse as a JSON file. parser throws an exception on errors
         // if successful, temporarily serialize to bytearray for communication
@@ -204,7 +205,8 @@
     if (comm->me == 0) {
       fp = fopen(filename.c_str(), "r");
       if (fp == nullptr)
-        error->one(FLERR, fileiarg, "Cannot open molecule file {}: {}", filename, utils::getsyserror());
+        error->one(FLERR, fileiarg, "Cannot open molecule file {}: {}", filename,
+                   utils::getsyserror());
     }
 
     // scan file for sizes of all fields and allocate storage for them
@@ -2345,6 +2347,12 @@
                (keyword == "Bond Coeffs") || (keyword == "Angle Coeffs") ||
                (keyword == "Dihedral Coeffs") || (keyword == "Improper Coeffs")) {
       error->all(FLERR, fileiarg, "Found data file section '{}' in molecule file\n", keyword);
+    } else if (keyword == "Representative atoms") { // read rep_atom info if provided
+      repatomflag = 1;
+      if (flag)
+        repatoms(line);
+      else
+        skip_lines(natoms,line,keyword);
     } else {
 
       // Error: Either a too long/short section or a typo in the keyword
@@ -3615,6 +3623,43 @@
 }
 
 /* ----------------------------------------------------------------------
+   read representative atoms from file
+------------------------------------------------------------------------- */
+
+void Molecule::repatoms(char *line) { // PP
+
+  const std::string location = "Representative atoms section of molecule file";
+  std::string repatomstr;
+  for (int i = 0; i < natoms; i++) count[i] = 0;
+
+  for (int i = 0; i < natoms; i++) {
+    readline(line);
+    auto values = Tokenizer(utils::trim(line)).as_vector();
+    int nwords = values.size();
+    
+    for (std::size_t ii = 0; ii < values.size(); ++ii) {
+      if (utils::strmatch(values[ii], "^#")) {
+        nwords = ii;
+        break;
+      }
+    }
+
+    if (nwords != 2) error->all(FLERR, "Invalid format in {}: {}", location, utils::trim(line)); 
+
+    // check # of atoms
+    int iatom = utils::inumeric(FLERR, values[0], false, lmp);
+    if (iatom < 1 || iatom > natoms)
+      error->all(FLERR, "Invalid atom index {} in {}: {}", iatom, location, utils::trim(line)); // should not happen
+    count[--iatom]++;
+
+    repatomstr = utils::utf8_subst(values[1]); // PP ... marked and unmarked atoms
+    rep_atom[iatom] = utils::inumeric(FLERR, repatomstr, false, lmp);
+  }
+}
+
+
+
+/* ----------------------------------------------------------------------
    error check molecule attributes and topology against system settings
 ------------------------------------------------------------------------- */
 
@@ -3698,6 +3743,7 @@
   nspecialflag = specialflag = 0;
   shakeflag = shakeflagflag = shakeatomflag = shaketypeflag = 0;
   bodyflag = ibodyflag = dbodyflag = 0;
+  repatomflag = 0;
 
   centerflag = massflag = comflag = inertiaflag = 0;
   tag_require = 0;
@@ -3759,6 +3805,7 @@
   if (muflag) memory->create(mu, natoms, 3, "molecule:mu");
   if (radiusflag) memory->create(radius, natoms, "molecule:radius");
   if (rmassflag) memory->create(rmass, natoms, "molecule:rmass");
+  if (repatomflag) memory->create(rep_atom, natoms, "molecule:rep_atom");
 
   // always allocate num_bond,num_angle,etc and nspecial
   // even if not in molecule file, initialize to 0
